File: C++/208. Implement Trie (Prefix Tree).cpp
#include <bits/stdc++.h>

using namespace std;

struct TrieNode
{
    vector<TrieNode *> map;
    bool is_end;

    TrieNode()
    {
        map = vector<TrieNode *>(26, nullptr);
        is_end = false;
    }
};

class Trie
{
    TrieNode *root;

public:
    Trie()
    {
        root = new TrieNode();
    }

    /** Inserts a word into the trie. */
    void insert(string word)
    {
        TrieNode *cur = root;
        for (char character : word)
        {
            character -= 'a';
            if (cur->map[character] == nullptr)
                cur->map[character] = new TrieNode();
            cur = cur->map[character]; // Move to next character
        }
        cur->is_end = true;
    }

    /** Returns if the word is in the trie. */
    bool search(string word)
    {
        TrieNode *cur = root;

        for (char character : word)
        {
            character -= 'a';
            if (cur->map[character] == nullptr)
                return false;
            cur = cur->map[character];
        }

        return cur->is_end;
    }

    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix)
    {
        TrieNode *cur = root;

        for (char character : prefix)
        {
            character -= 'a';
            if (cur->map[character] == nullptr)
                return false;
            cur = cur->map[character];
        }

        return true;
    }
};

File: Java/208. Implement Trie (Prefix Tree).java
class Trie {

    private final Node root;

    public Trie() {
        root = new Node();
    }

    public void insert(String word) {
        Node cur = root;
        for (Character ch : word.toCharArray()) {
            if (cur.characters[ch - 'a'] == null)
                cur.characters[ch - 'a'] = new Node();
            cur = cur.characters[ch - 'a'];
        }
        cur.isWord = true;
    }

    public boolean search(String word) {
        Node cur = root;
        for (Character ch : word.toCharArray()) {
            if (cur.characters[ch - 'a'] == null)
                return false;
            cur = cur.characters[ch - 'a'];
        }
        return cur.isWord;
    }

    public boolean startsWith(String prefix) {
        Node cur = root;
        for (Character ch : prefix.toCharArray()) {
            if (cur.characters[ch - 'a'] == null)
                return false;
            cur = cur.characters[ch - 'a'];
        }
        return true;
    }

    public class Node {
        public Node[] characters;
        public boolean isWord;

        public Node() {
            characters = new Node[26];
            isWord = false;
        }
    }
}